src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/services/RestrictedPatientCleanup.kt

```diff
@@ -25,17 +33,102 @@ class RestrictedPatientCleanup(
-        null
+        null,
       )
     } ?: log.debug("Movement into prison for prisoner {} ignored as not a restricted patient", prisonerNumber)
   }
 
-  fun mergeRestrictedPatient(removedNomsNumber: String?, nomsNumber: String) {
-    // TODO: Add in logic for merging restricted patient
-    log.debug("Merge event received - old prisoner number {}, new prisoner number {}", removedNomsNumber, nomsNumber)
+  fun mergeRestrictedPatient(removedPrisonerNumber: String?, prisonerNumber: String) {
+    if (!mergesEnabled) {
+      log.debug(
+        "Merge event disabled - old prisoner number {}, new prisoner number {}",
+        removedPrisonerNumber,
+        prisonerNumber,
+      )
+      return
+    }
+
+    restrictedPatientsRepository.findByIdOrNull(removedPrisonerNumber)?.let { removed ->
+      restrictedPatientsRepository.findByIdOrNull(prisonerNumber)?.let {
+        deleteRemovedRestrictedPatient(removed, it)
+      } ?: transferRemovedRestrictedPatient(removed, prisonerNumber)
+    } ?: log.debug("Merge for removed prisoner {} ignored as not a restricted patient", removedPrisonerNumber)
+  }
+
+  private fun deleteRemovedRestrictedPatient(removed: RestrictedPatient, kept: RestrictedPatient) {
+    // Both prisoner numbers were restricted patients. Check that they were at the same hospital before removing
+    if (removed.hospitalLocationCode == kept.hospitalLocationCode) {
+      restrictedPatientsRepository.delete(removed)
+      telemetryClient.trackEvent(
+        "restricted-patient-merged-removed-deleted",
+        mapOf("removedPrisonerNumber" to removed.prisonerNumber, "prisonerNumber" to kept.prisonerNumber),
+        null,
+      )
+    } else {
+      // different hospital so can't tell what the actual prison should be
+      throw MergeRestrictedPatientConflict(
+        "Patient is at two different hospitals during merge - removed ${removed.prisonerNumber} is at ${removed.hospitalLocationCode} vs ${kept.prisonerNumber} is at ${kept.hospitalLocationCode}"
+      )
+    }
+  }
+
+  private fun transferRemovedRestrictedPatient(patient: RestrictedPatient, keptPrisonerNumber: String) {
+    validateMergedPrisoner(keptPrisonerNumber, patient)
+
+    restrictedPatientsRepository.save(
+      RestrictedPatient(
+        prisonerNumber = keptPrisonerNumber,
+        fromLocationId = patient.fromLocationId,
+        hospitalLocationCode = patient.hospitalLocationCode,
+        supportingPrisonId = patient.supportingPrisonId,
+        dischargeTime = patient.dischargeTime,
+        commentText = patient.commentText,
+      )
+    )
+    restrictedPatientsRepository.delete(patient)
+    telemetryClient.trackEvent(
+      "restricted-patient-merged-removed-moved",
+      mapOf("removedPrisonerNumber" to patient.prisonerNumber, "prisonerNumber" to keptPrisonerNumber),
+      null,
+    )
+  }
+
+  private fun validateMergedPrisoner(keptPrisonerNumber: String, patient: RestrictedPatient) {
+    // TODO: implement validation
```

Comment:
since merges disabled in production, doesn't matter that I haven't finishing writing this yet 😁 .  Thought I'd keep PR small for now